File: demo/src/pages/StateExample/StateExample.swift

import Foundation
import SwiftUI

struct StateContentView: View {
    @StateObject private var state: ContentViewState = .init()
    var body: some View {
        VStack {
            Text(state.count.description)
            Button("Count Up") { state.countUp() }
            Button("Reset") { state.reset() }
        }
        .onReceive(state.$count) { count in
            print(count)
        }
    }
}

final class ContentViewState: ObservableObject {
    @Published private(set) var count: Int = 0
    
    func countUp() {
        count += 1
    }
    
    func reset() {
        count = 0
    }
}

File: demo/src/pages/StateExample/StateExample.ts
import { Bound, Button, ObservableObject, print, Published, StateObject, Text, Viewable, VStack } from "@tswift/ui";

export class StateExample extends Viewable {
  @StateObject state: ContentViewState = new ContentViewState();

  body = ({ state, $state }: Bound<this>) =>VStack(
    Text(state.count + ''),
    Button("Count Up", state.countUp),
    Button("Reset", state.reset)
  )
    .onReceive($state,  ()=>
      print(state.count))

}

class ContentViewState extends ObservableObject {
  @Published count: number = 0

  countUp = ()=> {
    this.count += 1
  }

  reset = ()=> {
    this.count = 0
  }
}
